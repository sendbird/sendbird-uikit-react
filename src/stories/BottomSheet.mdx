import { Meta, Canvas, ArgsTable } from '@storybook/addon-docs';
import BottomSheet from '../ui/BottomSheet';
import * as BottomSheetStories from './BottomSheet.stories';

<Meta title="UI/BottomSheet" component={BottomSheet} of={BottomSheetStories} />

# BottomSheet

The `BottomSheet` component is designed to present content in a sliding panel from the bottom of the screen, commonly used for menus, selecting options, or displaying information in a contained space that does not take over the entire screen. This component is highly effective for mobile-responsive designs and enhances user interaction within the application.

<Canvas of={BottomSheetStories.Default} />
<ArgsTable of={BottomSheet} />

### Importing BottomSheet

To incorporate the BottomSheet component into your project, use the following import statement:

```tsx
import BottomSheet from '@sendbird/uikit-react/ui/BottomSheet';
```
