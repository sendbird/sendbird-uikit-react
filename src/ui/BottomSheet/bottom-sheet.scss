@import '../../styles/variables';

.sendbird-bottomsheet {
  position: fixed;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
}

.sendbird-bottomsheet__content {
  position: fixed;
  z-index: 10001;
  bottom: 0;
  left: 0;
}

.sendbird-bottomsheet__backdrop {
  position: fixed;
  z-index: 10000;
  height: 100%;
  width: 100%;
  top: 0;
  left: 0;
  @include themed() {
    background-color: t(overlay-2);
  }
}
