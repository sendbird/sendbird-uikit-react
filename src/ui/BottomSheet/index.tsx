import './bottom-sheet.scss';
import React, { useRef } from 'react';
import { createPortal } from 'react-dom';

import { MODAL_ROOT } from '../../hooks/useModal/ModalRoot';

interface BottomSheetProps {
  className?: string;
  children: React.ReactElement;
  onBackdropClick?: () => void;
}

const BottomSheet: React.FunctionComponent<BottomSheetProps> = (props: BottomSheetProps) => {
  const {
    className = '',
    children,
    onBackdropClick,
  } = props;

  // https://github.com/testing-library/react-testing-library/issues/62#issuecomment-438653348
  const portalRoot = useRef<HTMLElement>();
  portalRoot.current = document.getElementById(MODAL_ROOT);
  if (!portalRoot.current) {
    portalRoot.current = document.createElement('div')
    portalRoot.current.setAttribute('id', MODAL_ROOT);
    document.body.appendChild(portalRoot.current);
  }
  return createPortal(
    <div
      className={`${className} sendbird-bottomsheet`}
    >
      <div
        className={'sendbird-bottomsheet__content'}
        role='dialog'
        aria-modal='true'
        aria-expanded='true'
      >
        {children}
      </div>
      <div
        className={`
          sendbird-bottomsheet__backdrop
        `}
        onClick={(e) =>{
          e?.stopPropagation();
          onBackdropClick();
        }}
      />
    </div>
  , portalRoot.current);
};

export default BottomSheet;
