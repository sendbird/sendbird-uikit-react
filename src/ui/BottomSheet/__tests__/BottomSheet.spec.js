import React from 'react';
import { render, screen } from '@testing-library/react';

import BottomSheet from "../index";

describe('ui/BottomSheet', () => {
  it('should do a snapshot test of the default Button DOM', function () {
    render(<BottomSheet className='test_classname' />);
    expect(document.body.lastChild).toMatchSnapshot();
  });
});
