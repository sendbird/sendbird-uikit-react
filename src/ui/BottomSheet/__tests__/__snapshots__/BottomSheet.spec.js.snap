// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/BottomSheet should do a snapshot test of the default Button DOM 1`] = `
<div
  id="sendbird-modal-root"
>
  <div
    class="test_classname sendbird-bottomsheet"
  >
    <div
      aria-expanded="true"
      aria-modal="true"
      class="sendbird-bottomsheet__content"
      role="dialog"
    />
    <div
      class="
          sendbird-bottomsheet__backdrop
        "
    />
  </div>
</div>
`;
